lib/Doctrine/ODM/MongoDB/DocumentManager.php

```diff
@@ -531,17 +531,33 @@ public function getRepository($documentName)
      * This effectively synchronizes the in-memory state of managed objects with the
      * database.
      *
-     * @param object $document
+     * @param object|array $documentOrOptions
      * @param array $options Array of options to be used with batchInsert(), update() and remove()
      * @throws \InvalidArgumentException
      */
-    public function flush($document = null, array $options = array())
+    public function flush($documentOrOptions = null, array $options = array())
     {
-        if (null !== $document && ! is_object($document) && ! is_array($document)) {
-            throw new \InvalidArgumentException(gettype($document));
+        if (func_num_args() === 1 && is_array($documentOrOptions) && ! $this->arrayContainsDocuments($documentOrOptions)) {
+            $options = $documentOrOptions;
+            $documentOrOptions = null;
+        }
+        if (null !== $documentOrOptions && ! is_object($documentOrOptions) && ! is_array($documentOrOptions)) {
+            throw new \InvalidArgumentException(gettype($documentOrOptions));
+        }
+        if ($documentOrOptions !== null) {
+            @trigger_error(
+                'Flushing selected documents has been deprecated and will be removed in doctrine/mongodb-odm 2.0.',
+                E_USER_DEPRECATED
+            );
+        }
+        if (func_num_args() === 2) {
```

Comment:
We should only trigger the deprecation notice if documents were passed:

```suggestion
        elseif (func_num_args() === 2) {
```

Reply:
Not necessarily, `$dm->flush(null, $options);` will stop working as well and folks can call `$dm->flush($options);` now

Reply:
As is, `$dm->flush([$document], $options);` throws two deprecation warnings. We should make sure this doesn't happen as it might confuse the user.

As I write this, we should also make sure to gracefully handle `$dm->flush([], $options);` to make sure we don't treat it as the "user passed documents" case.

Reply:
It seems that passing an array of documents is an undocumented feature. I didn't find anything in the documentation and the method signature mentions just _one_ object. Only [deep down in the UoW](https://github.com/doctrine/mongodb-odm/blob/1.3.x/lib/Doctrine/ODM/MongoDB/UnitOfWork.php#L404) is the check if the parameter is an array. So did anyone even use that?

Reply:
Hard to say, but there's a chance :)